File: bin/Data/LuaScripts/Utilities/2D/Mover.lua
-- Mover script object class
--    - Handles entity (enemy, platform...) translation along a path (set of Vector2 points)
--    - Supports looping paths and animation flip
--    - Default speed is 0.8 if 'Speed' property is not set in the tmx file objects

Mover = ScriptObject()

function Mover:Start()
    self.speed = 0.8
    self.path = {}
    self.currentPathID = 2
    self.emitTime = 0
    self.fightTimer = 0
    self.flip = 0
end

function Mover:Load(deserializer)
    self:SetPathAttr(deserializer:ReadBuffer())
end

function Mover:Save(serializer)
    serializer:WriteBuffer(self:GetPathAttr())
end

function Mover:SetPathAttr(buffer)
    if buffer.size == 0 then
        return
    end

    while not buffer.eof do
        table.insert(self.path, buffer:ReadVector2())
    end
end

function Mover:GetPathAttr()
    local ret = VectorBuffer()

    for i=1, table.maxn(self.path) do
        ret:WriteVector2(self.path[i])
    end

    return ret
end

function Mover:Update(timeStep)
    if table.maxn(self.path) < 2 then
        return
    end

    local node = self.node

    -- Handle Orc states (idle/wounded/fighting)
    if node.name == "Orc" then
        local animatedSprite = node:GetComponent("AnimatedSprite2D")
        local anim = "run"

        -- Handle wounded state
        if self.emitTime > 0 then
            self.emitTime = self.emitTime + timeStep
            anim = "dead"

            -- Handle dead
            if self.emitTime >= 3 then
                self.node:Remove()
                return
            end
        else
            -- Handle fighting state
            if self.fightTimer > 0 then
                anim = "attack"
                self.flip = character2DNode.position.x - node.position.x
                self.fightTimer = self.fightTimer + timeStep
                if self.fightTimer >= 3 then
                    self.fightTimer = 0 -- Reset
                end
            end
            -- Flip Orc animation according to speed, or player position when fighting
            animatedSprite.flipX = self.flip >= 0
        end
        -- Animate
        if animatedSprite.animation ~= anim then
            animatedSprite:SetAnimation(anim)
        end
    end

    -- Don't move if fighting or wounded
    if self.fightTimer > 0 or self.emitTime > 0 then
        return
    end

    -- Set direction and move to target
    local dir = self.path[self.currentPathID] - node.position2D
    local dirNormal = dir:Normalized()
    node:Translate(Vector3(dirNormal.x, dirNormal.y, 0) * Abs(self.speed) * timeStep)
    self.flip = dir.x

    if Abs(dir:Length()) < 0.1 then
        if self.speed > 0 then
            if self.currentPathID + 1 <= table.maxn(self.path) then
                self.currentPathID = self.currentPathID + 1
            else
                -- If loop, go to first waypoint, which equates to last one (and never reverse)
                if self.path[self.currentPathID] == self.path[1] then
                    self.currentPathID = 1
                    return
                end
                -- Reverse path if not looping
                self.currentPathID = self.currentPathID - 1
                self.speed = -self.speed
            end
        else
            if self.currentPathID - 1 > 0 then
                self.currentPathID = self.currentPathID - 1
            else
                -- Reverse path
                self.currentPathID = 2
                self.speed = -self.speed
            end
        end
    end
end

File: Source/Samples/Utilities2D/Mover.cpp
// Copyright (c) 2008-2022 the Urho3D project
// License: MIT

#include <Urho3D/Urho2D/AnimatedSprite2D.h>
#include <Urho3D/Urho2D/AnimationSet2D.h>
#include <Urho3D/Core/Context.h>
#include <Urho3D/IO/MemoryBuffer.h>
#include <Urho3D/Scene/Scene.h>
#include <Urho3D/Scene/SceneEvents.h>

#include <Urho3D/DebugNew.h>

#include "Mover.h"

Mover::Mover(Context* context) :
    LogicComponent(context),
    speed_(0.8f),
    currentPathID_(1),
    emitTime_(0.0f),
    fightTimer_(0.0f),
    flip_(0.0f)
{
    // Only the scene update event is needed: unsubscribe from the rest for optimization
    SetUpdateEventMask(LogicComponentEvents::Update);
}

void Mover::RegisterObject(Context* context)
{
    context->RegisterFactory<Mover>();

    // These macros register the class attribute to the Context for automatic load / save handling.
    // We specify the Default attribute mode which means it will be used both for saving into file, and network replication.
    URHO3D_ACCESSOR_ATTRIBUTE("Path", GetPathAttr, SetPathAttr, Variant::emptyBuffer, AM_DEFAULT);
    URHO3D_ATTRIBUTE("Speed", speed_, 0.8f, AM_DEFAULT);
    URHO3D_ATTRIBUTE("Current Path ID", currentPathID_, 1, AM_DEFAULT);
    URHO3D_ATTRIBUTE("Emit Time", emitTime_, 0.0f, AM_DEFAULT);
    URHO3D_ATTRIBUTE("Fight Timer", fightTimer_, 0.0f, AM_DEFAULT);
    URHO3D_ATTRIBUTE("Flip Animation", flip_, 0.0f, AM_DEFAULT);
}

void Mover::SetPathAttr(const Vector<byte>& value)
{
    if (value.Empty())
        return;

    MemoryBuffer buffer(value);
    while (!buffer.IsEof())
        path_.Push(buffer.ReadVector2());
}

Vector<byte> Mover::GetPathAttr() const
{
    VectorBuffer buffer;

    for (const Vector2& point : path_)
        buffer.WriteVector2(point);

    return buffer.GetBuffer();
}

void Mover::Update(float timeStep)
{
    if (path_.Size() < 2)
        return;

    // Handle Orc states (idle/wounded/fighting)
    if (node_->GetName() == "Orc")
    {
        auto* animatedSprite = node_->GetComponent<AnimatedSprite2D>();
        String anim = "run";

        // Handle wounded state
        if (emitTime_ > 0.0f)
        {
            emitTime_ += timeStep;
            anim = "dead";

            // Handle dead
            if (emitTime_ >= 3.0f)
            {
                node_->Remove();
                return;
            }
        }
        else
        {
            // Handle fighting state
            if (fightTimer_ > 0.0f)
            {
                anim = "attack";
                flip_ = GetScene()->GetChild("Imp", true)->GetPosition().x_ - node_->GetPosition().x_;
                fightTimer_ += timeStep;
                if (fightTimer_ >= 3.0f)
                    fightTimer_ = 0.0f; // Reset
            }
            // Flip Orc animation according to speed, or player position when fighting
            animatedSprite->SetFlipX(flip_ >= 0.0f);
        }
        // Animate
        if (animatedSprite->GetAnimation() != anim)
            animatedSprite->SetAnimation(anim);
    }

    // Don't move if fighting or wounded
    if (fightTimer_ > 0.0f || emitTime_ > 0.0f)
        return;

    // Set direction and move to target
    Vector2 dir = path_[currentPathID_] - node_->GetPosition2D();
    Vector2 dirNormal = dir.Normalized();
    node_->Translate(Vector3(dirNormal.x_, dirNormal.y_, 0.0f) * Abs(speed_) * timeStep);
    flip_ = dir.x_;

    // Check for new target to reach
    if (Abs(dir.Length()) < 0.1f)
    {
        if (speed_ > 0.0f)
        {
            if (currentPathID_ + 1 < path_.Size())
                currentPathID_ = currentPathID_ + 1;
            else
            {
                // If loop, go to first waypoint, which equates to last one (and never reverse)
                if (path_[currentPathID_] == path_[0])
                {
                    currentPathID_ = 1;
                    return;
                }
                // Reverse path if not looping
                currentPathID_ = currentPathID_ - 1;
                speed_ = -speed_;
            }
        }
        else
        {
            if (currentPathID_ - 1 >= 0)
                currentPathID_ = currentPathID_ - 1;
            else
            {
                currentPathID_ = 1;
                speed_ = -speed_;
            }
        }
    }
}
